core/src/main/java/io/questdb/cairo/O3Utils.java

```diff
@@ -25,26 +25,54 @@
 package io.questdb.cairo;
 
 import io.questdb.MessageBus;
+import io.questdb.cairo.sql.SqlExecutionCircuitBreakerConfiguration;
+import io.questdb.cairo.sql.async.PageFrameDispatchJob;
+import io.questdb.cairo.sql.async.PageFrameReduceJob;
 import io.questdb.log.Log;
 import io.questdb.log.LogFactory;
+import io.questdb.mp.Job;
 import io.questdb.mp.WorkerPool;
-import io.questdb.std.FilesFacade;
-import io.questdb.std.MemoryTag;
-import io.questdb.std.Unsafe;
-import io.questdb.std.Vect;
+import io.questdb.std.*;
+import org.jetbrains.annotations.Nullable;
 
 public class O3Utils {
 
     private static final Log LOG = LogFactory.getLog(O3Utils.class);
 
-    public static void setupWorkerPool(WorkerPool workerPool, MessageBus messageBus) {
-        O3PurgeDiscoveryJob purgeDiscoveryJob = new O3PurgeDiscoveryJob(messageBus, workerPool.getWorkerCount());
+    public static void setupWorkerPool(
+            WorkerPool workerPool,
+            MessageBus messageBus,
+            @Nullable SqlExecutionCircuitBreakerConfiguration sqlExecutionCircuitBreakerConfiguration
+    ) {
+        final int workerCount = workerPool.getWorkerCount();
+        final O3PurgeDiscoveryJob purgeDiscoveryJob = new O3PurgeDiscoveryJob(messageBus, workerPool.getWorkerCount());
+        final PageFrameDispatchJob pageFrameDispatchJob = new PageFrameDispatchJob(
+                messageBus,
+                workerCount,
+                sqlExecutionCircuitBreakerConfiguration
+        );
+
         workerPool.assign(purgeDiscoveryJob);
         workerPool.assign(new O3PartitionJob(messageBus));
         workerPool.assign(new O3OpenColumnJob(messageBus));
         workerPool.assign(new O3CopyJob(messageBus));
         workerPool.assign(new O3CallbackJob(messageBus));
+        workerPool.assign(pageFrameDispatchJob);
         workerPool.freeOnHalt(purgeDiscoveryJob);
+        workerPool.freeOnHalt(pageFrameDispatchJob);
+
+        for (int i = 0; i < workerCount; i++) {
+            // create job per worker to allow each worker to have
+            // own shard walk sequence
+            final PageFrameReduceJob pageFrameReduceJob = new PageFrameReduceJob(
+                    messageBus,
+                    new Rnd(),
```

Comment:
If we want each walk sequence to be different, we shouldn't use the same random seed in each of the jobs. We should either initialize each `Rnd` with different seeds or use `CairoConfiguration#getRandom`.

Reply:
really good point

Reply:
fixed here: https://github.com/questdb/questdb/pull/1732/commits/3b1973ab49fbe54a8c4c9d1ae49bef70815e253a